torch/amp/grad_scaling.py

```diff
@@ -0,0 +1,382 @@
+            as opposed to dynamic gradient scaling) and you're fairly
+            certain no inf/nan gradients will be encountered, setting ``skip_if_inf=False`` can provide a minor
+            performance improvement, because for typical optimizers it avoids copying ``found_inf``'s value back to
+            the CPU.  Another case where it makes sense to supply ``skip_if_inf=False`` is batch replay, where you
+            replay this iteration until inf/nan-free gradients are created.  By the time you reach the
+            ``optimizer.step_after_unscale()`` call, you're sure the gradients do not contain inf/nan, so it's safe
+            to elide the check.  For other (typical) use, the default ``skip_if_inf=True`` is strongly recommended.
+        scaling_enabled (bool, optional):  If ``False``, :func:`step_after_unscale` simply calls ``self.step()``
+            and returns ``opt_ret, None, None``, where opt_ret is the default return value of ``self.step()``.
+
+    Returns:
+        If ``found_inf`` contains 0.0, :func:`step_after_unscale` returns the default return value of the
+        ``self.step()`` call.  If ``found_inf`` contains > 0, :func:`step_after_unscale` returns None.
+    """
+    if not scaling_enabled:
+        return self.step(closure)
+
+    if closure is not None:
+        raise ValueError("Closure use is not currently supported.  It's tricky, but not impossible, and we're trying "
+                         "to decide if it's worth implementing.  If you require closure use, please comment on "
+                         "https://github.com/pytorch/pytorch/issues/25081, which will help us gauge demand.")
+
+    if skip_if_inf:
+        if found_inf.item():
+            return None
+        else:
+            return self.step()
+    else:
+        return self.step()
+
+
+def _add_amp_attributes(optimizer):
+    if not hasattr(optimizer, "unscale_and_step"):
+        optimizer.unscale_and_step = types.MethodType(unscale_and_step, optimizer)
+
+    if not hasattr(optimizer, "step_after_unscale"):
+        optimizer.step_after_unscale = types.MethodType(step_after_unscale, optimizer)
+
+    if not hasattr(optimizer, "unscale"):
+        optimizer.unscale = types.MethodType(unscale, optimizer)
+
+    if not hasattr(optimizer, "check_inf"):
+        optimizer.check_inf = types.MethodType(check_inf, optimizer)
+
+
+def add_amp_attributes(optimizers):
+    r"""
+    Adds convenience methods for gradient unscaling and inf/nan checking to an optimizer instance, and returns
+    a Tensor containing the recommended initial gradient scale (which you are free to use, or substitute your own).
+
+    The methods added are :func:`unscale`, :func:`check_inf`, :func:`unscale_and_step`, and :func`step_after_unscale`.
+
+    If an optimizer some or all of these methods defined already, :func:`add_amp_attributes` won't touch them, allowing
```

Comment:
I think this note can be elsewhere for optimizer implementers. Also, per the point above, it seems like if optimizers are implementing these functions on their own then they should just implement them all (possibly by explicitly importing the 'generic' versions) to avoid confusion.